File: tests/src/test/cpp/userauth/UserAuthenticationDataTest.cpp
/**
 * Copyright (c) 2016-2017 Netflix, Inc.  All rights reserved.
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <gtest/gtest.h>
#include <gmock/gmock.h>
#include <userauth/UserAuthenticationData.h>
#include <entityauth/EntityAuthenticationScheme.h>
#include <keyx/KeyExchangeScheme.h>
#include <userauth/EmailPasswordAuthenticationFactory.h>
#include <userauth/EmailPasswordStore.h>
#include <util/AuthenticationUtils.h>
#include <MslEncodingException.h>
#include <MslUserAuthException.h>
#include <userauth/UserAuthenticationFactory.h>
#include <memory>
#include <string>

#include "../util/MockMslContext.h"

using namespace std;
using namespace testing;
using namespace netflix::msl;
using namespace netflix::msl::entityauth;
using namespace netflix::msl::io;
using namespace netflix::msl::tokens;
using namespace netflix::msl::util;

namespace netflix {
namespace msl {
namespace userauth {

namespace {

/** Key user authentication scheme. */
const string KEY_SCHEME = "scheme";
/** Key user authentication data. */
const string KEY_AUTHDATA = "authdata";

class MockUserAuthenticationFactory : public UserAuthenticationFactory
{
    MockUserAuthenticationFactory() : UserAuthenticationFactory(UserAuthenticationScheme::EMAIL_PASSWORD) {}

    MOCK_METHOD3(createData, std::shared_ptr<UserAuthenticationData>(std::shared_ptr<util::MslContext> ctx,
            std::shared_ptr<tokens::MasterToken> masterToken, std::shared_ptr<io::MslObject> userAuthMo));

    MOCK_METHOD4(authenticate, std::shared_ptr<tokens::MslUser>(std::shared_ptr<util::MslContext> ctx,
            const std::string& identity, std::shared_ptr<UserAuthenticationData> data,
            std::shared_ptr<tokens::UserIdToken> userIdToken));
};

// FIXME: duplicated in EmailPasswordAuthenticationData_test
class MockEmailPasswordStore : public EmailPasswordStore
{
    MOCK_METHOD2(isUser, std::shared_ptr<tokens::MslUser>(const std::string& email, const std::string& password));
};

// FIXME: duplicated in EmailPasswordAuthenticationData_test
class MockAuthenticationUtils : public AuthenticationUtils
{
    MOCK_METHOD1(isEntityRevoked, bool(const std::string& identity));
    MOCK_METHOD2(isSchemePermitted, bool(const std::string& identity, const entityauth::EntityAuthenticationScheme& scheme));
    MOCK_METHOD2(isSchemePermitted, bool(const std::string& identity, const userauth::UserAuthenticationScheme& scheme));
    MOCK_METHOD3(isSchemePermitted, bool(const std::string& identity, shared_ptr<tokens::MslUser> user, const userauth::UserAuthenticationScheme& scheme));
    MOCK_METHOD2(isSchemePermitted, bool(const std::string& identity, const keyx::KeyExchangeScheme& scheme));
};

}  // namespace anonymous

/**
 * User authentication data unit tests.
 *
 * Successful calls to
 * {@link UserAuthenticationData#create(com.netflix.msl.util.MslContext, org.json.MslObject)}
 * covered in the individual user authentication data unit tests.
 */
class UserAuthenticationDataTest : public ::testing::Test
{
public:
    UserAuthenticationDataTest()
    : ctx(make_shared<MockMslContext>(EntityAuthenticationScheme::PSK, false))
    , encoder(ctx->getMslEncoderFactory())
    {}
protected:
    shared_ptr<MockMslContext> ctx;
    shared_ptr<MslEncoderFactory> encoder;
};

TEST_F(UserAuthenticationDataTest, noScheme)
{
//    thrown.expect(MslEncodingException.class);
//    thrown.expectMslError(MslError.MSL_PARSE_ERROR);

    shared_ptr<MslObject> mo = encoder->createObject();
    mo->put(KEY_SCHEME + "x", UserAuthenticationScheme::EMAIL_PASSWORD.name());
    mo->put(KEY_AUTHDATA, encoder->createObject());
    try {
        UserAuthenticationData::create(ctx, shared_ptr<MasterToken>(), mo);
        ADD_FAILURE() << "Should have thrown";
    } catch (const MslEncodingException& e) {
        EXPECT_EQ(MslError::MSL_PARSE_ERROR, e.getError());
    }
}

TEST_F(UserAuthenticationDataTest, noAuthdata)
{
//    thrown.expect(MslEncodingException.class);
//    thrown.expectMslError(MslError.MSL_PARSE_ERROR);

    shared_ptr<MslObject> mo = encoder->createObject();
    mo->put(KEY_SCHEME, UserAuthenticationScheme::EMAIL_PASSWORD.name());
    mo->put(KEY_AUTHDATA + "x", encoder->createObject());
    try {
        UserAuthenticationData::create(ctx, shared_ptr<MasterToken>(), mo);
        ADD_FAILURE() << "Should have thrown";
    } catch (const MslEncodingException& e) {
        EXPECT_EQ(MslError::MSL_PARSE_ERROR, e.getError());
    }
}

TEST_F(UserAuthenticationDataTest, unidentifiedScheme)
{
//    thrown.expect(MslUserAuthException.class);
//    thrown.expectMslError(MslError.UNIDENTIFIED_USERAUTH_SCHEME);

    shared_ptr<MslObject> mo = encoder->createObject();
    const string invalidSchemeName = "foobar";
    mo->put<string>(KEY_SCHEME, invalidSchemeName);
    mo->put(KEY_AUTHDATA, encoder->createObject());
    try {
        UserAuthenticationData::create(ctx, shared_ptr<MasterToken>(), mo);
        ADD_FAILURE() << "Should have thrown";
    } catch (const MslUserAuthException& e) {
        EXPECT_EQ(MslError::UNIDENTIFIED_USERAUTH_SCHEME, e.getError());
    }
}

TEST_F(UserAuthenticationDataTest, authFactoryNotFound)
{
//    thrown.expect(MslUserAuthException.class);
//    thrown.expectMslError(MslError.USERAUTH_FACTORY_NOT_FOUND);

	shared_ptr<MockMslContext> ctx = make_shared<MockMslContext>(EntityAuthenticationScheme::PSK, false);
	ctx->removeUserAuthenticationFactory(UserAuthenticationScheme::EMAIL_PASSWORD);

    shared_ptr<MslObject> mo = encoder->createObject();
    mo->put(KEY_SCHEME, UserAuthenticationScheme::EMAIL_PASSWORD.name());
    mo->put(KEY_AUTHDATA, encoder->createObject());
    try {
        UserAuthenticationData::create(ctx, shared_ptr<MasterToken>(), mo);
        ADD_FAILURE() << "Should have thrown";
    } catch (const MslUserAuthException& e) {
        EXPECT_EQ(MslError::USERAUTH_FACTORY_NOT_FOUND, e.getError());
    }
}

}}} // namespace netflix::msl::userauth

File: tests/src/test/javascript/userauth/UserAuthenticationDataTest.js
/**
 * Copyright (c) 2012-2017 Netflix, Inc.  All rights reserved.
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * User authentication data unit tests.
 * 
 * Successful calls to
 * {@link UserAuthenticationData#create(com.netflix.msl.util.MslContext, org.json.JSONObject)}
 * covered in the individual user authentication data unit tests.
 * 
 * @author Wesley Miaw <wmiaw@netflix.com>
 */
describe("UserAuthenticationData", function() {
    var EntityAuthenticationScheme = require('msl-core/entityauth/EntityAuthenticationScheme.js');
    var UserAuthenticationScheme = require('msl-core/userauth/UserAuthenticationScheme.js');
    var UserAuthenticationData = require('msl-core/userauth/UserAuthenticationData.js');
    var MslEncodingException = require('msl-core/MslEncodingException.js');
    var MslError = require('msl-core/MslError.js');
    var MslUserAuthException = require('msl-core/MslUserAuthException.js');

    var MslTestConstants = require('msl-tests/MslTestConstants.js');
    var MockMslContext = require('msl-tests/util/MockMslContext.js');

    /** Key user authentication scheme. */
    var KEY_SCHEME = "scheme";
    /** Key user authentication data. */
    var KEY_AUTHDATA = "authdata";

    /** MSL context. */
    var ctx;
    /** MSL encoder factory. */
    var encoder;

    var initialized = false;
    beforeEach(function() {
        if (!initialized) {
            runs(function() {
                MockMslContext.create(EntityAuthenticationScheme.PSK, false, {
                    result: function(c) { ctx = c; },
                    error: function(e) { expect(function() { throw e; }).not.toThrow(); }
                });
            });
            waitsFor(function() { return ctx; }, "ctx", MslTestConstants.TIMEOUT_CTX);
            runs(function() {
                encoder = ctx.getMslEncoderFactory();
                initialized = true;
            });
        }
    });

    it("no scheme", function() {
        var exception;
        runs(function() {
            var mo = encoder.createObject();
            mo.put(KEY_SCHEME + "x", UserAuthenticationScheme.EMAIL_PASSWORD.name);
            mo.put(KEY_AUTHDATA, encoder.createObject());
            UserAuthenticationData.parse(ctx, null, mo, {
                result: function() {},
                error: function(e) { exception = e; }
            });
        });
        waitsFor(function() { return exception; }, "exception", MslTestConstants.TIMEOUT);

        runs(function() {
            var f = function() { throw exception; };
            expect(f).toThrow(new MslEncodingException(MslError.MSL_PARSE_ERROR));
        });
    });

    it("no authdata", function() {
        var exception;
        runs(function() {
            var mo = encoder.createObject();
            mo.put(KEY_SCHEME, UserAuthenticationScheme.EMAIL_PASSWORD.name);
            mo.put(KEY_AUTHDATA + "x", encoder.createObject());
            UserAuthenticationData.parse(ctx, null, mo, {
                result: function() {},
                error: function(e) { exception = e; }
            });
        });
        waitsFor(function() { return exception; }, "exception", MslTestConstants.TIMEOUT);

        runs(function() {
            var f = function() { throw exception; };
            expect(f).toThrow(new MslEncodingException(MslError.MSL_PARSE_ERROR));
        });
    });

    it("unidentified scheme", function() {
        var exception;
        runs(function() {
            var mo = encoder.createObject();
            mo.put(KEY_SCHEME, "x");
            mo.put(KEY_AUTHDATA, encoder.createObject());
            UserAuthenticationData.parse(ctx, null, mo, {
                result: function() {},
                error: function(e) { exception = e; }
            });
        });
        waitsFor(function() { return exception; }, "exception", MslTestConstants.TIMEOUT);

        runs(function() {
            var f = function() { throw exception; };
            expect(f).toThrow(new MslUserAuthException(MslError.UNIDENTIFIED_USERAUTH_SCHEME));
        });
    });

    it("authentication factory not found", function() {
        var ctx;
        runs(function() {
            MockMslContext.create(EntityAuthenticationScheme.PSK, false, {
                result: function(c) { ctx = c; },
                error: function(e) { expect(function() { throw e; }).not.toThrow(); }
            });
        });
        waitsFor(function() { return ctx; }, "ctx", MslTestConstants.TIMEOUT);

        var exception;
        runs(function() {
            ctx.removeUserAuthenticationFactory(UserAuthenticationScheme.EMAIL_PASSWORD);
            var mo = encoder.createObject();
            mo.put(KEY_SCHEME, UserAuthenticationScheme.EMAIL_PASSWORD.name);
            mo.put(KEY_AUTHDATA, encoder.createObject());
            UserAuthenticationData.parse(ctx, null, mo, {
                result: function() {},
                error: function(e) { exception = e; }
            });
        });
        waitsFor(function() { return exception; }, "exception", MslTestConstants.TIMEOUT);

        runs(function() {
            var f = function() { throw exception; };
            expect(f).toThrow(new MslUserAuthException(MslError.USERAUTH_FACTORY_NOT_FOUND));
        });
    });
});